ddl/ddl.go

```diff
@@ -376,21 +390,49 @@ func (d *ddl) Start(ctxPool *pools.ResourcePool) error {
 			return errors.Trace(err)
 		}
 
-		d.workers = make(map[workerType]*worker, 2)
 		d.sessPool = newSessionPool(ctxPool)
 		d.delRangeMgr = d.newDeleteRangeManager(ctxPool == nil)
-		d.workers[generalWorker] = newWorker(d.ctx, generalWorker, d.sessPool, d.delRangeMgr)
-		d.workers[addIdxWorker] = newWorker(d.ctx, addIdxWorker, d.sessPool, d.delRangeMgr)
-		for _, worker := range d.workers {
-			worker.wg.Add(1)
-			w := worker
-			go w.start(d.ddlCtx)
 
-			metrics.DDLCounter.WithLabelValues(fmt.Sprintf("%s_%s", metrics.CreateDDL, worker.String())).Inc()
+		if AllowConcurrentDDL.Load() {
+			addIdxWorkerFunc := func() (pools.Resource, error) {
+				wk := newWorker(d.ctx, addIdxWorker, d.sessPool, d.delRangeMgr)
 
-			// When the start function is called, we will send a fake job to let worker
-			// checks owner firstly and try to find whether a job exists and run.
-			asyncNotify(worker.ddlJobCh)
+				metrics.DDLCounter.WithLabelValues(fmt.Sprintf("%s_%s", metrics.CreateDDL, wk.String())).Inc()
+
+				// When the start function is called, we will send a fake job to let worker
+				// checks owner firstly and try to find whether a job exists and run.
+				asyncNotify(wk.ddlJobCh)
+				return wk, nil
+			}
+			generalWorkerFunc := func() (pools.Resource, error) {
+				wk := newWorker(d.ctx, generalWorker, d.sessPool, d.delRangeMgr)
+
+				metrics.DDLCounter.WithLabelValues(fmt.Sprintf("%s_%s", metrics.CreateDDL, wk.String())).Inc()
+
+				// When the start function is called, we will send a fake job to let worker
+				// checks owner firstly and try to find whether a job exists and run.
+				asyncNotify(wk.ddlJobCh)
+				return wk, nil
+			}
+			d.addIdxWorker = newDDLWorkerPool(pools.NewResourcePool(addIdxWorkerFunc, 10, 10, 3*time.Minute))
+			d.generalWorker = newDDLWorkerPool(pools.NewResourcePool(generalWorkerFunc, 10, 10, 0))
+			d.sessForAddDDL, _ = d.sessPool.get()
+			d.wg.Run(d.startDispatchLoop)
+		} else {
+			d.workers = make(map[workerType]*worker, 2)
```

Comment:
we have worker pool, no need these two workers now

Reply:
just keep the old DDL code style. but I think it will be improved in the future.